Fix SIL function side effects to handle unapplied escaping captures.

atrick opened the pull request:
Fix SIL function side effects to handle unapplied escaping

Fixes rdar://113339972 DeadStoreElimination causes uninitialized closure context

Before this fix, the recently enabled function side effect implementation
would return no side effects for a partial apply that is not applied
in the same function. This resulted in DeadStoreElimination
incorrectly eliminating the initialization of the closure context.

The fix is to model the effects of capturing the arguments for the
closure context. The effects of running the closure body will be
considered later, at the point that the closure is
applied. Running the closure does, however, depend on the captured
values to be valid. If the value being captured is addressible,
then we need to model the effect of reading from that memory. In
this case, the capture reads from a local stack slot:

    %stack = alloc_stack $Klass
    store %ref to %stack : $*Klass
    %closure = partial_apply [callee_guaranteed] %f(%stack)
      : $@convention(thin) (@in_guaranteed Klass) -> ()

Later, when the closure is applied, we won't have any reference back
to the original stack slot. The application may not even happen in a caller.

Note that, even if the closure will be applied in the current
function, the side effects of the application are insufficient to
cover the side effects of the capture. For example, the closure
body itself may not read from an argument, but the context must
still be valid in case it is copied or if the capture itself was
not a bitwise-move.

As an optimization, we ignore the effect of captures for on-stack
partial applies. Such captures are always either a bitwise-move
or, more commonly, capture the source value by address. In these
cases, the side effects of applying the closure are sufficient to
cover the effects of the captures. And, if an on-stack closure is
not invoked in the current function (or passed to a callee) then
it will never be invoked, so the captures never have effects.

atrick commented:
@swift-ci test

atrick commented:
@swift-ci benchmark

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ComputeSideEffects.swift:
```diff
@@ -142,6 +142,15 @@ private struct CollectedEffects {
+      // Nonescaping closures may bitwise-move value into the context,
+      // but this has no effect.
+      if !pa.isOnStack {
+        for operand in pa.operands {
```
We don't need to include the first operand (= the callee). For this you can write
```
  for argument in pa.arguments {
```

atrick replied:
@eeckstein I intended to handle the callee argument in case the callee is itself an escaping closure which might have been passed in. I need to add back a comment about that and come up with a SIL test case.

eeckstein reviewed (approved):
lgtm, thanks

atrick committed:
Fix SIL function side effects to handle unapplied escaping

Fixes rdar://113339972 DeadStoreElimination causes uninitialized closure context

Before this fix, the recently enabled function side effect implementation
would return no side effects for a partial apply that is not applied
in the same function. This resulted in DeadStoreElimination
incorrectly eliminating the initialization of the closure context.

The fix is to model the effects of capturing the arguments for the
closure context. The effects of running the closure body will be
considered later, at the point that the closure is
applied. Running the closure does, however, depend on the captured
values to be valid. If the value being captured is addressible,
then we need to model the effect of reading from that memory. In
this case, the capture reads from a local stack slot:

    %stack = alloc_stack $Klass
    store %ref to %stack : $*Klass
    %closure = partial_apply [callee_guaranteed] %f(%stack)
      : $@convention(thin) (@in_guaranteed Klass) -> ()

Later, when the closure is applied, we won't have any reference back
to the original stack slot. The application may not even happen in a caller.

Note that, even if the closure will be applied in the current
function, the side effects of the application are insufficient to
cover the side effects of the capture. For example, the closure
body itself may not read from an argument, but the context must
still be valid in case it is copied or if the capture itself was
not a bitwise-move.

As an optimization, we ignore the effect of captures for on-stack
partial applies. Such captures are always either a bitwise-move
or, more commonly, capture the source value by address. In these
cases, the side effects of applying the closure are sufficient to
cover the effects of the captures. And, if an on-stack closure is
not invoked in the current function (or passed to a callee) then
it will never be invoked, so the captures never have effects.

atrick commented:
@swift-ci test

atrick commented:
@swift-ci benchmark

atrick commented:
No benchmark changes.

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ComputeSideEffects.swift:
```diff
@@ -142,6 +142,24 @@ private struct CollectedEffects {
+      if !pa.isOnStack {
+        // the callee and its arguments are all captured...
+        for operand in pa.operands {
+          if operand.value.type.isAddress {
```
@atrick For the record: the callee (operand 0) never has an address type. So you could write `for argument in pa.arguments`. It's just cosmetic, but would make it clear that only arguments can contribute to the read-effects.

atrick replied:
@eeckstein I didn't think the callee could be an address but also didn't see anything in the SILVerifier prohibiting it. The closure result must be an object of course, but it says nothing about the captured callee. I suppose we never reabstract over function types in a way that requires an address, so I'll just add an assert to SILVerifier and fix this code as you suggested.
